libs/contrib/Text/Lexer/Core.idr

```diff
@@ -0,0 +1,190 @@
+module Text.Lexer.Core
+
+import public Control.Delayed
+
+%default total
+
+||| A language of token recognisers.
+||| @ consumes If `True`, this recogniser is guaranteed to consume at
+|||            least one character of input when it succeeds.
+export
+data Recognise : (consumes : Bool) -> Type where
+     Empty : Recognise False
+     Fail : Recognise c
+     Lookahead : (positive : Bool) -> Recognise c -> Recognise False
+     Pred : (Char -> Bool) -> Recognise True
+     SeqEat : Recognise True -> Inf (Recognise e) -> Recognise True
+     SeqEmpty : Recognise e1 -> Recognise e2 -> Recognise (e1 || e2)
+     Alt : Recognise e1 -> Recognise e2 -> Recognise (e1 && e2)
+
+||| A token recogniser. Guaranteed to consume at least one character.
+public export
+Lexer : Type
+Lexer = Recognise True
+
+||| Sequence two recognisers. If either consumes a character, the sequence
+||| is guaranteed to consume a character.
+export %inline
+(<+>) : {c1 : Bool} ->
+        Recognise c1 -> inf c1 (Recognise c2) -> Recognise (c1 || c2)
+(<+>) {c1 = False} = SeqEmpty
+(<+>) {c1 = True} = SeqEat
+
+||| Alternative recognisers. If both consume, the combination is guaranteed
+||| to consumer a character.
+export
+(<|>) : Recognise c1 -> Recognise c2 -> Recognise (c1 && c2)
+(<|>) = Alt
+
+||| A recogniser that always fails.
+export
+fail : Recognise c
+fail = Fail
+
+||| Recognise no input (doesn't consume any input)
+export
+empty : Recognise False
+empty = Empty
+
+||| Recognise a character that matches a predicate
+export
+pred : (Char -> Bool) -> Lexer
+pred = Pred
+
+||| Positive lookahead. Never consumes input.
+export
+expect : Recognise c -> Recognise False
+expect = Lookahead True
+
+||| Negative lookahead. Never consumes input.
+export
+reject : Recognise c -> Recognise False
+reject = Lookahead False
+
+||| Combine each lexer in a container into a lexer that sequences them.
+||| The resulting lexer always fails if the container was empty.
+export
+concat : Foldable t => t Lexer -> Lexer
+concat xs = foldl op Fail xs
+  where
+    op : Lexer -> Lexer -> Lexer
+    op Fail x = x
+    op x y = x <+> y
```

Comment:
Normally, `concat` and `choice` do the same thing, except that `concat` involves a `Type` and `choice` involves a `Type -> Type`. It happens that `Lexer` has two monoidal structures on it; the “additive” one with the operation being alternation and the identity element failure, and the “multiplicative” one with the operation being concatenation and the identity element being, not failure, but a lexer that always succeeds with no result.

Reply:
See [Wikipedia on Semiring of formal languages](https://en.wikipedia.org/wiki/Semiring#formal_languages). In the context of `Applicative` and `Alternative`, the multiplicative counterpart of `choice` would be `sequence_`/`sequence`.

Reply:
I did it this way so that the return value could be `Lexer`. Always succeeding would mean that the return type would have to be `Recognize False` instead, which isn't a problem except I will have to check that the strings are not empty in `exact` and `approx`.

Reply:
Actually, there is not a way to succeed if the foldable is empty, unless the return type is `(c : Bool ** Recognise c)`.

Reply:
There are at least two ways to fix it, but both require you take a `List` rather than any `Foldable`. One is using `NonEmpty`, and the other is specifying the return value using `isCons`.

Also, do you not want to be able to sequence a collection of `Recognise False`s?

Reply:
I went with the `isCons` variant. Both that and `NonEmpty` are somewhat cumbersome to use, but `isCons` is more flexible.